crates/environ/src/module.rs

```diff
@@ -261,6 +311,156 @@ impl ModuleTranslation<'_> {
+        // * Only initializers for defined memories.
+        // * Only initializers with static offsets (no globals).
+        // * Only in-bound initializers.
+        //
+        // The `init_memory` method of `MemoryInitialization` is used here to
+        // do most of the validation for us, and otherwise the data chunks are
+        // collected into the `images` array here.
+        let mut images: PrimaryMap<MemoryIndex, Vec<u8>> = PrimaryMap::default();
+        let num_defined_memories =
+            self.module.memory_plans.len() - self.module.num_imported_memories;
+        for _ in 0..num_defined_memories {
+            images.push(Vec::new());
+        }
+        let mut data = self.data.iter();
+        let ok = self.module.memory_initialization.init_memory(
+            InitMemory::CompileTime(&self.module),
+            &mut |memory, init| {
+                let data = data.next().unwrap();
+                assert_eq!(data.len(), init.data.len());
+
+                // Static initialization with only one memcpy is only possible
+                // for defined memories which have a known-starting-as-zero
+                // state to account for holes between data segments. This means
+                // that if this is an imported memory then static memory
+                // initialization isn't possible.
+                if self.module.defined_memory_index(memory).is_none() {
+                    return false;
+                }
+
+                // Splat the `data_range` into the `image` for this memory,
+                // updating it as necessary with 0s for holes and such.
+                let image = &mut images[memory];
+                let offset = usize::try_from(init.offset).unwrap();
+                let new_image_len = offset + data.len();
+                if image.len() < new_image_len {
+                    if new_image_len > MAX_IMAGE_SIZE {
+                        return false;
+                    }
+                    image.resize(new_image_len, 0);
+                }
+                image[offset..][..data.len()].copy_from_slice(data);
+                true
+            },
+        );
+
+        // If any initializer wasn't applicable then we skip static init
+        // entirely.
+        if !ok {
+            return;
+        }
+
+        // At this point all memories in this module are initialized with
+        // in-bounds initializers which are all statically known to be valid.
+        // This means that the memory `images` built up so far are valid for an
+        // instance of this linear memory. These images are trimmed of their
+        // leading and trailing zeros and then `self.data` is re-populated with
+        // new data.
+        self.data.clear();
+        assert!(self.data_align.is_none());
+        self.data_align = Some(page_size);
+        let mut map = PrimaryMap::with_capacity(images.len());
+        let mut offset = 0u32;
+        for (defined_memory, mut image) in images {
```

Comment:
here also the index variable `defined_memory` seems to imply that `images` only contains `DefinedMemoryIndex`s; should be consistent with what we do above. `map` contains an entry for every memory, defined or imported, so this loop is correct if `images` is over all memories as well; just a naming issue I think.

Reply:
Ah yeah this is a historical name, good catch though and definitely needed a rename.